File: 백준/Bronze/9506. 약수들의 합/약수들의 합.cc
#include <algorithm>
#include <iostream>
#include <vector>

std::vector<int> findDivisors(int n) {
  std::vector<int> divisors;
  for (int i = 1; i * i <= n; i++) {
    if (n % i == 0) {
      divisors.push_back(i);
      if (i != n / i) {
        divisors.push_back(n / i);
      }
    }
  }
  std::sort(divisors.begin(), divisors.end());
  return divisors;
}

int main() {
  while (true) {
    int n;
    std::cin >> n;
    if (n == -1) {
      break;
    }
    std::vector<int> divisors = findDivisors(n);
    int sum = 0;
    for (auto &it : divisors) {
      if (it != n) {
        sum += it;
      }
    }

    if (sum == n) {
      std::cout << n << " = ";
      for (int i = 0; i < divisors.size() - 1; i++) {
        std::cout << divisors[i] << (i != divisors.size() - 2 ? " + " : "\n");
      }
    } else {
      std::cout << n << " is NOT perfect." << std::endl;
    }
  }
  return 0;
}

File: 백준/Bronze/9506. 약수들의 합/약수들의 합.py
import sys

def proper_divisors(n: int) -> list[int]:
    divisors = {1}
    for i in range(2, int(n ** 0.5) + 1):
        if n % i == 0:
            divisors.add(i)
            divisors.add(n // i)
    if n == 1:
        return []
    return sorted(divisors)

def main() -> None:
    outputs = []
    for line in sys.stdin:
        value = int(line)
        if value == -1:
            break
        divisors = proper_divisors(value)
        if sum(divisors) == value:
            decomposition = ' + '.join(map(str, divisors))
            outputs.append(f"{value} = {decomposition}")
        else:
            outputs.append(f"{value} is NOT perfect.")
    sys.stdout.write('
'.join(outputs))

if __name__ == '__main__':
    main()
